src/test/libcephfs/vxattr.cc

```diff
@@ -0,0 +1,203 @@
+#include <sys/xattr.h>
+#endif
+
+#include <fmt/format.h>
+#include <map>
+#include <vector>
+#include <thread>
+#include <regex>
+
+#ifndef ALLPERMS
+#define ALLPERMS (S_ISUID|S_ISGID|S_ISVTX|S_IRWXU|S_IRWXG|S_IRWXO)
+#endif
+
+using namespace std;
+
+TEST(LibCephFS, GetAndSetDirLayout) {
+
+  struct ceph_mount_info *cmount;
+  ASSERT_EQ(0, ceph_create(&cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_read_file(cmount, NULL));
+  ASSERT_EQ(0, ceph_conf_parse_env(cmount, NULL));
+  ASSERT_EQ(0, ceph_mount(cmount, "/"));
+
+  ASSERT_EQ(0, ceph_mkdirs(cmount, "test/d0/subdir", 0777));
+
+  char value[1024];
+  int r = 0;
+
+  // check for default layout
+  r = ceph_getxattr(cmount, "test/d0", "ceph.dir.layout", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  std::clog << "layout:" << value << std::endl;
+  ASSERT_STRNE((char*)NULL, strstr(value, "\"inheritance\": \"@inherited\""));
+
+  // set a new layout and verify the same
+  const char *new_layout = "{"
+			      "\"stripe_unit\": 65536, "
+			      "\"stripe_count\": 1, "
+			      "\"object_size\": 65536, "
+			      "\"pool_name\": \"cephfs.a.data\", "
+			   "}";
+  ASSERT_EQ(0, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout", (void*)new_layout, strlen(new_layout), XATTR_REPLACE));
+  r = ceph_getxattr(cmount, "test/d0", "ceph.dir.layout", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  std::clog << "layout:" << value << std::endl;
+  ASSERT_STRNE((char*)NULL, strstr(value, "\"inheritance\": \"@set\""));
+
+  // now check that the subdir layout is inherited
+  r = ceph_getxattr(cmount, "test/d0/subdir", "ceph.dir.layout", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+  std::clog << "layout:" << value << std::endl;
+  ASSERT_STRNE((char*)NULL, strstr(value, "\"inheritance\": \"@inherited\""));
+
+  // try to set a malformed JSON, eg. without an open brace
+  ASSERT_EQ(-EINVAL, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout", (void*)(&value[1]), r-1, XATTR_REPLACE));
+
+  // try setting a bad pool name
+  ASSERT_EQ(-EINVAL, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout.pool_name", (void*)"UglyPoolName", 12, XATTR_REPLACE));
+  r = ceph_getxattr(cmount, "test/d0", "ceph.dir.layout.pool_name", (void*)value, sizeof(value));
+  ASSERT_GT(r, 0);
+  value[r] = '\0';
+
+  // try setting a wrong pool name
+  ASSERT_EQ(-EINVAL, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout.pool_name", (void*)"cephfs.a.meta", 13, XATTR_REPLACE));
+
+  // try setting a bad pool id
+  ASSERT_EQ(-EINVAL, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout.pool_id", (void*)"300", 3, XATTR_REPLACE));
+
+  // try to set in invalid field
+  ASSERT_EQ(-ENODATA, ceph_setxattr(cmount, "test/d0", "ceph.dir.layout.bad_field", (void*)"300", 3, XATTR_REPLACE));
+
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test/d0/subdir"));
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test/d0"));
+  ASSERT_EQ(0, ceph_rmdir(cmount, "test"));
+
+  ceph_shutdown(cmount);
+}
```

Comment:
break this into separate tests:

* default layout
* new layout (set) + inherited subdir
* malformed json set
* bad pool name
* etc.

Reply:
ack; layout tests have now been split into separate ones